### PR TITLE
Migrate Junit4 to Junit5

### PR DESCRIPTION
**Pull request by stanlyDoge**
### Type of change
- Refactoring

### Description
Please review the topic operator tests with increased attention.

### Checklist

- [ ] Update/write design documentation in `./design`
- [ ] Write tests
- [ ] Make sure all tests pass
- [ ] Update documentation
- [ ] Check RBAC rights for Kubernetes / OpenShift roles
- [ ] Try your changes from Pod inside your Kubernetes and OpenShift cluster, not just locally
- [ ] Reference relevant issue(s) and close them after merging
- [ ] Update CHANGELOG.md

### CODE REVIEW

**Comment by stanlyDoge** [comment]
@strimzi-ci run tests

File: certificate-manager/src/test/java/io/strimzi/certs/SecretCertProviderTest.java [hunk 1]
```diff
@@ -58,15 +57,15 @@ public void testKeyAndCertInSecret() throws Exception {
...
-        assertEquals("123456", new String(decoder.decode(secret.getData().get("truststore.password"))));
+        assertThat(secret.getMetadata().getName(), is("my-secret"));
+        assertThat(secret.getMetadata().getNamespace(), is("my-namespace"));
+        assertThat(secret.getMetadata().getOwnerReferences().size(), is(1));
+        assertThat(secret.getMetadata().getOwnerReferences().get(0), is(ownerReference));
+        assertThat(secret.getData().size(), is(4));
+        assertThat(Arrays.equals(Files.readAllBytes(key.toPath()), decoder.decode(secret.getData().get("ca.key"))), is(true));
+        assertThat(Arrays.equals(Files.readAllBytes(cert.toPath()), decoder.decode(secret.getData().get("ca.crt"))), is(true));
+        assertThat(Arrays.equals(Files.readAllBytes(store.toPath()), decoder.decode(secret.getData().get("truststore.p12"))), is(true));
+        assertThat(new String(decoder.decode(secret.getData().get("truststore.password"))), is("123456"));
```
Comment (Seequick1) [inline certificate-manager/src/test/java/io/strimzi/certs/SecretCertProviderTest.java]: Should we create here some method to remove redundancy? Same code from line 99.

File: cluster-operator/src/test/java/io/strimzi/operator/cluster/ClusterOperatorConfigTest.java [hunk 2]
```diff
@@ -44,29 +45,29 @@ public void testDefaultConfig() {
...
     public void testReconciliationInterval() {
 
         ClusterOperatorConfig config = new ClusterOperatorConfig(singleton("namespace"), 60_000, 30_000, false, new KafkaVersion.Lookup(emptyMap(), emptyMap(), emptyMap(), emptyMap()), null, null);
 
-        assertEquals(singleton("namespace"), config.getNamespaces());
-        assertEquals(60_000, config.getReconciliationIntervalMs());
-        assertEquals(30_000, config.getOperationTimeoutMs());
+        assertThat(config.getNamespaces(), is(singleton("namespace")));
+        assertThat(config.getReconciliationIntervalMs(), is(60_000L));
+        assertThat(config.getOperationTimeoutMs(), is(30_000L));
```
Comment (Seequick1) [inline cluster-operator/src/test/java/io/strimzi/operator/cluster/ClusterOperatorConfigTest.java]: Wouldn't be better have some Constants which express `defaultOperatiorTimeMs`? What do you think ? Same with `defaultReconcilationIntervalMs`

File: cluster-operator/src/test/java/io/strimzi/operator/cluster/model/AbstractConfigurationTest.java [hunk 3]
```diff
@@ -261,9 +261,9 @@ public void testKafkaMirrorMakerConsumerHostnameVerification() {
 
         KafkaMirrorMakerConsumerConfiguration configuration = new KafkaMirrorMakerConsumerConfiguration(conf.entrySet());
 
-        assertEquals("zstd", configuration.asOrderedProperties().asMap().get("compression.type"));
-        assertNull(configuration.asOrderedProperties().asMap().get("ssl.keystore.location"));
-        assertEquals("", configuration.asOrderedProperties().asMap().get("ssl.endpoint.identification.algorithm"));
+        assertThat(configuration.asOrderedProperties().asMap().get("compression.type"), is("zstd"));
+        assertThat(configuration.asOrderedProperties().asMap().get("ssl.keystore.location"), is(nullValue()));
+        assertThat(configuration.asOrderedProperties().asMap().get("ssl.endpoint.identification.algorithm"), is(""));
```
Comment (Seequick1) [inline cluster-operator/src/test/java/io/strimzi/operator/cluster/model/AbstractConfigurationTest.java]: here also create some generic method like `verifyComponentConfigurations`. Line `250 - 252, 264 - 266, 278 - 280`

File: cluster-operator/src/test/java/io/strimzi/operator/cluster/model/AbstractModelTest.java [hunk 4]
```diff
@@ -71,28 +71,28 @@ protected String getDefaultLogConfigFileName() {
     public void testJvmMemoryOptionsXmsOnly() {
         Map<String, String> env = getStringStringMap(null, "4",
                 0.5, 5_000_000_000L, null);
-        assertEquals("-Xms4", env.get(AbstractModel.ENV_VAR_KAFKA_HEAP_OPTS));
-        assertEquals(null, env.get(AbstractModel.ENV_VAR_DYNAMIC_HEAP_FRACTION));
-        assertEquals(null, env.get(AbstractModel.ENV_VAR_DYNAMIC_HEAP_MAX));
+        assertThat(env.get(AbstractModel.ENV_VAR_KAFKA_HEAP_OPTS), is("-Xms4"));
+        assertThat(env.get(AbstractModel.ENV_VAR_DYNAMIC_HEAP_FRACTION), is(nullValue()));
+        assertThat(env.get(AbstractModel.ENV_VAR_DYNAMIC_HEAP_MAX), is(nullValue()));
```
Comment (Seequick1) [inline cluster-operator/src/test/java/io/strimzi/operator/cluster/model/AbstractModelTest.java]: same here

File: cluster-operator/src/test/java/io/strimzi/operator/cluster/model/EntityTopicOperatorTest.java [hunk 5]
```diff
@@ -104,33 +104,33 @@
...
+        assertThat(entityTopicOperator.livenessProbeOptions.getPeriodSeconds(), is(livenessProbe.getPeriodSeconds()));
+        assertThat(entityTopicOperator.getWatchedNamespace(), is(toWatchedNamespace));
+        assertThat(entityTopicOperator.getReconciliationIntervalMs(), is(toReconciliationInterval * 1000));
+        assertThat(entityTopicOperator.getZookeeperSessionTimeoutMs(), is(toZookeeperSessionTimeout * 1000));
+        assertThat(entityTopicOperator.getZookeeperConnect(), is(EntityTopicOperator.defaultZookeeperConnect(cluster)));
+        assertThat(entityTopicOperator.getKafkaBootstrapServers(), is(EntityTopicOperator.defaultBootstrapServers(cluster)));
+        assertThat(entityTopicOperator.getResourceLabels(), is(ModelUtils.defaultResourceLabels(cluster)));
+        assertThat(entityTopicOperator.getTopicMetadataMaxAttempts(), is(toTopicMetadataMaxAttempts));
+        assertThat(entityTopicOperator.getLogging().getType(), is(topicOperatorLogging.getType()));
+        assertThat(((InlineLogging) entityTopicOperator.getLogging()).getLoggers(), is(topicOperatorLogging.getLoggers()));
```
Comment (Seequick1) [inline cluster-operator/src/test/java/io/strimzi/operator/cluster/model/EntityTopicOperatorTest.java]: EntityTopicOperatorTest, EntityUserOperatorTest and etc. can also has some base method like `verifyComponentCrds` or name it more meaningfull

File: cluster-operator/src/test/java/io/strimzi/operator/cluster/model/KafkaClusterTest.java [hunk 6]
```diff
@@ -166,22 +162,22 @@ private void checkMetricsConfigMap(ConfigMap metricsCm) {
...
+        assertThat(headful.getSpec().getPorts().get(1).getName(), is(KafkaCluster.CLIENT_PORT_NAME));
+        assertThat(headful.getSpec().getPorts().get(1).getPort(), is(new Integer(KafkaCluster.CLIENT_PORT)));
+        assertThat(headful.getSpec().getPorts().get(1).getProtocol(), is("TCP"));
+        assertThat(headful.getSpec().getPorts().get(2).getName(), is(KafkaCluster.CLIENT_TLS_PORT_NAME));
+        assertThat(headful.getSpec().getPorts().get(2).getPort(), is(new Integer(KafkaCluster.CLIENT_TLS_PORT)));
+        assertThat(headful.getSpec().getPorts().get(2).getProtocol(), is("TCP"));
+        assertThat(headful.getSpec().getPorts().get(3).getName(), is(AbstractModel.METRICS_PORT_NAME));
+        assertThat(headful.getSpec().getPorts().get(3).getPort(), is(new Integer(KafkaCluster.METRICS_PORT)));
+        assertThat(headful.getSpec().getPorts().get(3).getProtocol(), is("TCP"));
+        assertThat(headful.getMetadata().getAnnotations(), is(kc.getPrometheusAnnotations()));
```
Comment (Seequick1) [inline cluster-operator/src/test/java/io/strimzi/operator/cluster/model/KafkaClusterTest.java]: same here `verifyServiceProperties` or something like that

File: cluster-operator/src/test/java/io/strimzi/operator/cluster/model/KafkaConnectClusterTest.java [hunk 7]
```diff
@@ -219,51 +215,55 @@ public void testGenerateServiceWithoutMetrics()   {
...
+        assertThat(dep.getSpec().getTemplate().getSpec().getContainers().get(0).getPorts().size(), is(2));
+        assertThat(dep.getSpec().getTemplate().getSpec().getContainers().get(0).getPorts().get(0).getContainerPort(), is(new Integer(KafkaConnectCluster.REST_API_PORT)));
+        assertThat(dep.getSpec().getTemplate().getSpec().getContainers().get(0).getPorts().get(0).getName(), is(KafkaConnectCluster.REST_API_PORT_NAME));
+        assertThat(dep.getSpec().getTemplate().getSpec().getContainers().get(0).getPorts().get(0).getProtocol(), is("TCP"));
+        assertThat(dep.getSpec().getStrategy().getType(), is("RollingUpdate"));
+        assertThat(dep.getSpec().getStrategy().getRollingUpdate().getMaxSurge().getIntVal(), is(new Integer(1)));
+        assertThat(dep.getSpec().getStrategy().getRollingUpdate().getMaxUnavailable().getIntVal(), is(new Integer(0)));
+        assertThat(AbstractModel.containerEnvVars(dep.getSpec().getTemplate().getSpec().getContainers().get(0)).get(KafkaConnectCluster.ENV_VAR_KAFKA_CONNECT_TLS), is(nullValue()));
```
Comment (Seequick1) [inline cluster-operator/src/test/java/io/strimzi/operator/cluster/model/KafkaConnectClusterTest.java]: Are we able to create some method for verifying deployment ?

File: cluster-operator/src/test/java/io/strimzi/operator/cluster/model/KafkaConnectS2IClusterTest.java [hunk 8]
```diff
@@ -151,51 +145,51 @@ private void checkMetricsConfigMap(ConfigMap metricsCm) {
...
+        assertThat(kc.image, is(KafkaConnectS2IResources.deploymentName(cluster) + ":latest"));
+        assertThat(kc.replicas, is(replicas));
+        assertThat(kc.sourceImageBaseName + ":" + kc.sourceImageTag, is(image));
+        assertThat(kc.readinessProbeOptions.getInitialDelaySeconds(), is(healthDelay));
+        assertThat(kc.readinessProbeOptions.getTimeoutSeconds(), is(healthTimeout));
+        assertThat(kc.livenessProbeOptions.getInitialDelaySeconds(), is(healthDelay));
+        assertThat(kc.livenessProbeOptions.getTimeoutSeconds(), is(healthTimeout));
+        assertThat(kc.getConfiguration().asOrderedProperties(), is(expectedConfiguration));
+        assertThat(kc.bootstrapServers, is(bootstrapServers));
+        assertThat(kc.isInsecureSourceRepository(), is(false));
```
Comment (Seequick1) [inline cluster-operator/src/test/java/io/strimzi/operator/cluster/model/KafkaConnectS2IClusterTest.java]: same

File: cluster-operator/src/test/java/io/strimzi/operator/cluster/model/KafkaConnectS2IClusterTest.java [hunk 9]
```diff
@@ -151,51 +145,51 @@ private void checkMetricsConfigMap(ConfigMap metricsCm) {
...
+        assertThat(svc.getMetadata().getLabels(), is(expectedLabels(KafkaConnectS2IResources.serviceName(cluster))));
+        assertThat(svc.getSpec().getSelector(), is(expectedSelectorLabels()));
+        assertThat(svc.getSpec().getPorts().size(), is(2));
+        assertThat(svc.getSpec().getPorts().get(0).getPort(), is(new Integer(KafkaConnectCluster.REST_API_PORT)));
+        assertThat(svc.getSpec().getPorts().get(0).getName(), is(KafkaConnectCluster.REST_API_PORT_NAME));
+        assertThat(svc.getSpec().getPorts().get(0).getProtocol(), is("TCP"));
+        assertThat(svc.getSpec().getPorts().get(1).getName(), is(AbstractModel.METRICS_PORT_NAME));
+        assertThat(svc.getSpec().getPorts().get(1).getPort(), is(new Integer(KafkaCluster.METRICS_PORT)));
+        assertThat(svc.getSpec().getPorts().get(1).getProtocol(), is("TCP"));
+        assertThat(svc.getMetadata().getAnnotations(), is(kc.getPrometheusAnnotations()));
```
Comment (Seequick1) [inline cluster-operator/src/test/java/io/strimzi/operator/cluster/model/KafkaConnectS2IClusterTest.java]: here you can use that genertic method for verifying service configuration

File: cluster-operator/src/test/java/io/strimzi/operator/cluster/model/KafkaConnectS2IClusterTest.java [hunk 10]
```diff
@@ -229,80 +223,82 @@ public void testGenerateServiceWithoutMetrics()   {
...
+        assertThat(dep.getSpec().getTriggers().get(1).getImageChangeParams().getAutomatic(), is(true));
+        assertThat(dep.getSpec().getTriggers().get(1).getImageChangeParams().getContainerNames().size(), is(1));
+        assertThat(dep.getSpec().getTriggers().get(1).getImageChangeParams().getContainerNames().get(0), is(KafkaConnectS2IResources.deploymentName(this.cluster)));
+        assertThat(dep.getSpec().getTriggers().get(1).getImageChangeParams().getFrom().getName(), is(KafkaConnectS2IResources.deploymentName(this.cluster) + ":latest"));
+        assertThat(dep.getSpec().getTriggers().get(1).getImageChangeParams().getFrom().getKind(), is("ImageStreamTag"));
+        assertThat(dep.getSpec().getStrategy().getType(), is("Rolling"));
+        assertThat(dep.getSpec().getStrategy().getRollingParams().getMaxSurge().getIntVal(), is(new Integer(1)));
+        assertThat(dep.getSpec().getStrategy().getRollingParams().getMaxUnavailable().getIntVal(), is(new Integer(0)));
+        assertThat(AbstractModel.containerEnvVars(dep.getSpec().getTemplate().getSpec().getContainers().get(0)).get(KafkaConnectCluster.ENV_VAR_KAFKA_CONNECT_TLS),
+                is(nullValue()));
```
Comment (Seequick1) [inline cluster-operator/src/test/java/io/strimzi/operator/cluster/model/KafkaConnectS2IClusterTest.java]: also here deployment method

File: cluster-operator/src/test/java/io/strimzi/operator/cluster/model/KafkaMirrorMakerClusterTest.java [hunk 11]
```diff
@@ -176,40 +172,38 @@ public void testDefaultValues() {
...
-        assertEquals("TCP", svc.getSpec().getPorts().get(0).getProtocol());
-        assertEquals(mm.getPrometheusAnnotations(), svc.getMetadata().getAnnotations());
+        assertThat(svc.getSpec().getType(), is("ClusterIP"));
+        assertThat(svc.getMetadata().getLabels(), is(expectedLabels(mm.getServiceName())));
+        assertThat(svc.getSpec().getSelector(), is(expectedSelectorLabels()));
+        assertThat(svc.getSpec().getPorts().size(), is(1));
+        assertThat(svc.getSpec().getPorts().get(0).getName(), is(AbstractModel.METRICS_PORT_NAME));
+        assertThat(svc.getSpec().getPorts().get(0).getPort(), is(new Integer(KafkaCluster.METRICS_PORT)));
+        assertThat(svc.getSpec().getPorts().get(0).getProtocol(), is("TCP"));
+        assertThat(svc.getMetadata().getAnnotations(), is(mm.getPrometheusAnnotations()));
```
Comment (Seequick1) [inline cluster-operator/src/test/java/io/strimzi/operator/cluster/model/KafkaMirrorMakerClusterTest.java]: same

**Comment by stanlyDoge** [comment]
@strimzi-ci run tests profile=acceptance

**Comment by stanlyDoge** [comment]
@strimzi-ci run tests

**Comment by stanlyDoge** [comment]
@Seequick1 as we discussed offline - I do not think this kind of refactoring should be part of this PR. Let's wait for what others think. Also, it would be great to have this merged ASAP because of resolving conflicts after each merged PR with tests.

**Comment by Frawless** [comment]
Regarding the system tests results,  I think we face the same problem as in https://github.com/strimzi/strimzi-kafka-operator/pull/2149

**Comment by scholzj** [comment]
@stanlyDoge You have to resolve the conflicts before we can merge it

**Pull request merged**